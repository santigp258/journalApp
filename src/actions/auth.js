import { types } from "../types/types";


//return login action
export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});
