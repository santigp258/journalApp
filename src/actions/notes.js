import Swal from "sweetalert2";

import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
  //async homework
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activeNote(docRef.id, newNote));
    dispatch(startLoadingNotes(uid));
  };
};

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

//save note in FireStore
export const startSaveNotes = (note) => {
  return async (dispatch, getState) => {
    //user uid
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFirestore);

    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved!", note.title, "success");
  };
};

//refresh one note
export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note, //no delete key
    },
  },
});

//upload file

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({
      title: "Uploading...",
      text: "Please Wait",
      willOpen: () => {
        Swal.showLoading();
      },
    });
    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    dispatch(startSaveNotes(activeNote));

    Swal.close();
  };
};

//delete from firestore
export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    await db.doc(`/${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

//delete note
export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const actionCleaning = () => ({
  type: types.notesActionCleaning,
});

export const noteLogout = () => ({
  type: types.notesLogoutCleaning,
});
