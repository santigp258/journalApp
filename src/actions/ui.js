import { types } from "../types/types";

//using implicit return () => ({})
export const setError = (err) => ({
  type: types.uiSetError,
  payload: err,
});

export const removeError = () => ({
  type: types.uiRemoveError,
});

export const startLoading = () => ({
  type: types.uiStartLoading,
});
export const finishLoading = () => ({
  type: types.uiFinishLoading,
});
