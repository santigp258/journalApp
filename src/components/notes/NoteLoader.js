import React from "react";

export const NoteLoader = () => {
  return (
    <div className="loader__container">
      <div className="loader">
        <div id="loader__Indicator">
          <div className="loader__Bar" id="loader__Bar1"></div>
          <div className="loader__Bar" id="loader__Bar2"></div>
          <div className="loader__Bar" id="loader__Bar3"></div>
          <div className="loader__Bar" id="loader__Bar4"></div>
        </div>
      </div>
    </div>
  );
};
