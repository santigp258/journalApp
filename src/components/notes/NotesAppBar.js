import React, { useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNotes, startUploading } from "../../actions/notes";
import moment from "moment";

export const NotesAppBar = () => {
  //time
  const currentTime = new Date().getTime();
  const noteDate = moment(currentTime);

  const dispatch = useDispatch();
  //notes state
  const { active } = useSelector((state) => state.notes);

  //save note
  const handleSave = () => {
    dispatch(startSaveNotes(active));
  };

  const inputUploadImage = useRef();
  const handlePictureUpload = () => {
    inputUploadImage.current.click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if (file) {
      dispatch(startUploading(file));
    }
  };
  return (
    <div className="notes__appbar">
      <span>{noteDate.format("dddd, MMMM Do")}</span>

      <input
        type="file"
        style={{ display: "none" }}
        onChange={handleFileChange}
        ref={inputUploadImage}
      />
      <div>
        <button className="btn" onClick={handlePictureUpload}>
          Picture
        </button>
        <button className="btn" onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};
