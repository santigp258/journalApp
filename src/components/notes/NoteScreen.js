import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNote } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
  const { active: note } = useSelector((state) => state.notes);
  const [formValues, handleInputChange, reset] = useForm(note);

  const dispatch = useDispatch();
  //active note
  const activeId = useRef(note.id);
  useEffect(() => {
    //show changes when one note selected
    if (note.id !== activeId.current) {
      reset(note, reset);

      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    //refresh active note
    dispatch(activeNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  const { body, title } = formValues;
  return (
    <div className="notes__main-content">
      <NotesAppBar />

      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="notes__title-input"
          name="title"
          onChange={handleInputChange}
          value={title}
          autoComplete="off"
        />

        <textarea
          placeholder="What Happened today?"
          className="notes__textarea"
          name="body"
          onChange={handleInputChange}
          value={body}
          autoComplete="off"
        ></textarea>
        <div className="notes__image">
          {note.url && <img src={note.url} alt="Stars" />}
        </div>
      </div>
    </div>
  );
};
