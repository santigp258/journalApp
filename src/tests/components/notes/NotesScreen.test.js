import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router";
import { NoteScreen } from "../../../components/notes/NoteScreen";
import { activeNote } from "../../../actions/notes";
const middlewares = [thunk];

jest.mock("../../../actions/notes", () => ({
  activeNote: jest.fn(),
}));
//create store
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: { loading: false, msgError: null },
  notes: {
    active: { id: "abc", title: "Hi", body: "mundo", date: 12344 },
    notes: [],
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn();
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <NoteScreen />
    </MemoryRouter>
  </Provider>
);

describe("pruebas en <NotesScreen/>", () => {
  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de disparar el active note", () => {
    wrapper.find('input[name="title"]').simulate("change", {
      target: {
        name: "title",
        value: "Hello again",
      },
    });

    expect(activeNote).toHaveBeenCalled()
  });
});
