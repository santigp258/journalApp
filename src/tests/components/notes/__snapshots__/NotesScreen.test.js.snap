// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <NotesScreen/> debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Saturday, March 27th
    </span>
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hi"
    />
    <textarea
      autoComplete="off"
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What Happened today?"
      value="mundo"
    />
    <div
      className="notes__image"
    />
  </div>
  <button
    className="btn btn-warning"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
